File: picasso/src/main/java/com/squareup/picasso3/SvgImageDecoder.java
package com.squareup.picasso3;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.util.Log;
import com.caverock.androidsvg.SVG;
import com.caverock.androidsvg.SVGParseException;
import java.io.IOException;
import okio.BufferedSource;

class SvgImageDecoder implements ImageDecoder {

  @Override public boolean canHandleSource(BufferedSource source) {
    SourceBufferingInputStream wrapped = new SourceBufferingInputStream(source);
    try {
      SVG svg = SVG.getFromInputStream(wrapped);
      return true;
    } catch (SVGParseException e) {
      Log.e("Test", "Failed to parse SVG: " + e.getMessage(), e);
      return false;
    }
  }

  @Override public Image decodeImage(BufferedSource source, Request request) throws IOException {
    SourceBufferingInputStream wrapped = new SourceBufferingInputStream(source);
    try {
      SVG svg = SVG.getFromInputStream(wrapped);
      if (request.hasSize()) {
        if (request.targetWidth != 0) {
          svg.setDocumentWidth(request.targetWidth);
        }
        if (request.targetHeight != 0) {
          svg.setDocumentHeight(request.targetHeight);
        }
      }

      final int width = (int) svg.getDocumentWidth();
      final int height = (int) svg.getDocumentHeight();
      Bitmap bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);
      Canvas canvas = new Canvas(bitmap);
      svg.renderToCanvas(canvas);

      return new Image(bitmap);
    } catch (SVGParseException e) {
      throw new IOException(e);
    }
  }
}

File: picasso/src/main/java/com/squareup/picasso3/SvgImageDecoder.kt
package com.squareup.picasso3

import android.graphics.Bitmap
import android.graphics.Canvas
import android.util.Log
import com.caverock.androidsvg.SVG
import com.caverock.androidsvg.SVGParseException
import okio.BufferedSource
import java.io.BufferedInputStream
import java.io.IOException


internal class SvgImageDecoder : ImageDecoder {
    override fun canHandleSource(source: BufferedSource?): Boolean {
        val source = source!!.peek()
        return try {
            SVG.getFromInputStream(BufferedInputStream(source.inputStream()))
            true
        } catch (e: SVGParseException) {
            Log.e("Test", "Failed to parse SVG: " + e.message, e)
            false
        }
    }

    @Throws(IOException::class)
    override fun decodeImage(source: BufferedSource?, request: Request?): ImageDecoder.Image? {
        val source = source!!.peek()
        return try {
            val svg: SVG = SVG.getFromInputStream(BufferedInputStream(source.inputStream()))
            if (request!!.hasSize()) {
                if (request.targetWidth != 0) {
                    svg.documentWidth = request.targetWidth.toFloat()
                    svg.documentHeight = request.targetHeight.toFloat()
                }
                if (request.targetHeight != 0) {
                    svg.documentHeight = request.targetHeight.toFloat()
                }
            }
            val width = svg.documentWidth.toInt()
            val height = svg.documentHeight.toInt()
            val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
            val canvas = Canvas(bitmap)
            svg.renderToCanvas(canvas)
            ImageDecoder.Image(bitmap)
        } catch (e: SVGParseException) {
            throw IOException(e)
        }
    }
}